src/operator/tensor/control_flow_op.h

```diff
@@ -0,0 +1,159 @@
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y) {
+    out[i] = (0 != cond[i]? x[i] : y[i]);
+  }
+};
+
+/*! \brief return elements from x or y depending on condition
+ * This is the specialized template for the operator where.
+ * The condition is a vector whose size is the same as the
+ * x's first dim size.
+ * The returned array is formed by rows from x or y depending on
+ * the condition's elements.
+ */
+template<>
+struct where<false> {
+  // DType is the output data type
+  // CType is the condition data type
+  template<typename DType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const CType* cond,
+                                  const DType* x, const DType* y, int M) {
+    int k = i * M;
+    const DType* a = (cond[i] != 0? x : y);
+    for (int j = 0; j < M; ++j) {
+        out[k] = a[k];
+        ++k;
+    }
+  }
+};
+
+inline bool WhereOpShape(const nnvm::NodeAttrs& attrs,
+                         std::vector<TShape>* in_attrs,
+                         std::vector<TShape>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3)
+    << "where operator takes 3 arguments (" << in_attrs->size() << " given)";
+  CHECK_EQ(out_attrs->size(), 1);
+
+  // condition shape
+  const TShape& cshape = (*in_attrs)[0];
+  // x shape
+  const TShape& xshape = (*in_attrs)[1];
+  // y shape
+  const TShape& yshape = (*in_attrs)[2];
+
+  CHECK_EQ(xshape, yshape) << "x must have the same shape as y";
+  CHECK_LE(cshape.ndim(), xshape.ndim())
+    << "condition's dimension=" << cshape.ndim()
+    << " cannot be greater than x's dimension=" << xshape.ndim();
+
+  // handle 0-dim arrays
+  if (cshape.ndim() == 0 && xshape.ndim() == 0) {
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape());
+    return true;
+  }
+
+  // condition, x, and y cannot be 0-dim arrays
+  // if the program has reached here
+  CHECK_GT(cshape.ndim(), 0) << "condition array cannot be 0-dim if x is not 0-dim";
+  CHECK_GT(xshape.ndim(), 0) << "x array cannot 0-dim if condition is not 0-dim";
+
+  // condition and x have the same shape
+  if (cshape.ndim() == xshape.ndim()) {
+    CHECK_EQ(cshape, xshape) << "condition and x have the same dimension but different shapes";
+  } else {
+    CHECK_EQ(cshape.ndim(), 1) << "condition must either have the same shape as x or be a vector";
+    CHECK_EQ(cshape[0], xshape[0]) << "condition's first dim size ("
+      << cshape[0] << ") must be equal to x's first dim size (" << xshape[0] << ")";
+  }
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, xshape);
+  return true;
+}
+
+inline bool WhereOpType(const nnvm::NodeAttrs& attrs,
+                        std::vector<int>* in_attrs,
+                        std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3)
+    << "where operator takes 3 arguments (" << in_attrs->size() << " given)";
+  CHECK_EQ(out_attrs->size(), 1);
+
+  if ((*in_attrs)[0] == -1) {
```

Comment:
see ElemwiseType, don't assign float32 as default. reduce dtype from in[1], in[2] and out[0]. in[0] can be left as-is